### ISSUE TITLE
Document stubbing services caveat

### ISSUE THREAD

**Issue by mehulkar** [ember-learn/guides-source#1262]
https://guides.emberjs.com/release/testing/testing-components/#toc_stubbing-services

This section shows how to stub a service in a component test, but there is a caveat. If a service was accessed in an initializer, calling `register` will not work. This should be documented somehow. A workaround would be nice as well, but I don't think there *is* a good workaround at the moment.

**Comment by rwjblue** [comment]
Confirm, there isn't a good work around. FWIW, I *do* think that it is plausible for us to error if the thing being registered has already been instantiated. I'm not sure exactly how often that would hit folks, but it at least would make it clearer when you are going to be hitting a problem...

**Comment by mehulkar** [comment]
I haven't tested this, but `this.owner.unregister` *does* work, but it's even more of a self-own because you can't unregister and re-register the mock until after initializers have run. So no your test runs *some* of the app code with the real service and the rest with the mock.

**Comment by rwjblue** [comment]
> I believe this.owner.unregister does work

"work" here is a bit relative, specifically calling `.unregister` will remove the cached singleton that was previously created however it cannot remove all references to the original value. So if you have a service that is injected into a route or other service, then you unregister/reregister, that original value will be cached by anything that looked it up prior to the unregister/register.

tldr; I don't think that is an acceptable level of working 😸

**Comment by mehulkar** [comment]
Oh my god, yes! I'm pretty sure I ran into this actually (but with a different registered singleton) and it was a nightmare to debug.
